_specifications/lsp/3.18/types/diagnostic.md

```diff
@@ -35,8 +39,11 @@ export interface Diagnostic {
 
 	/**
 	 * The diagnostic's message.
+	 * 
+	 * @since 3.18.0 - support for MarkupContent. This is guarded by the client
+	 * capability `workspace.diagnostic.markupMessageSupport`.
 	 */
-	message: string;
+	message: string | MarkupContent;
```

Comment:
It will be a breaking change (type-wise) if the type suddenly changes from `string` to `string | MarkupContent`. I think this might need to be a separate property.

Reply:
Realistically the servers will probably have to support both plain string and markup since there is no guarantee that all clients will support markup. So it might make sense for servers to provide both variants in separate properties (when client supports markup).

Reply:
> It will be a breaking change (type-wise) if the type suddenly changes from string to string | MarkupContent. I think this might need to be a separate property.

I think this is fine given that it is behind a capability. Note how previous proposals followed the same approach; e.g. [snipped edits](https://github.com/microsoft/language-server-protocol/pull/1892).

Reply:
> Realistically servers will probably have to support both plain string and markup since there is no guarantee that all clients will support markup. So it might make sense for servers to provide both variants in separate properties (when client supports markup).

True, although one could argue that this is also true for other features that support markup content, yet those don't provide both formats.

Reply:
That being said, @dbaeumer has the final call.